packages/app/src/components/DBTraceWaterfallChart.tsx

```diff
@@ -22,233 +27,338 @@ type SpanRow = {
+  logTableModel?: TSource;
   traceId: string;
   dateRange: [Date, Date];
   focusDate: Date;
-  onClick?: (rowWhere: string) => void;
+  onClick?: (rowWhere: { id: string; type: string }) => void;
   highlightedRowWhere?: string | null;
 }) {
-  const { data, isFetching, aliasMap } = useSpansAroundFocus({
-    traceTableModel,
+  const { rows: traceRowsData, isFetching: traceIsFetching } =
+    useSpansAroundFocus({
+      tableModel: traceTableModel,
+      focusDate,
+      dateRange,
+      traceId,
+    });
+  const { rows: logRowsData, isFetching: logIsFetching } = useSpansAroundFocus({
+    // search data if logTableModel exist
+    // search invliad date range if no logTableModel(react hook need execute no matter what)
+    tableModel: logTableModel || traceTableModel,
     focusDate,
-    dateRange,
+    dateRange: logTableModel ? dateRange : [dateRange[1], dateRange[0]],
     traceId,
   });
 
-  const rowWhere = useRowWhere({ meta: data?.meta, aliasMap });
-
-  const rows = useMemo(
-    () =>
-      data?.meta != null && data.meta.length > 0 // Sometimes meta has not loaded yet
-        ? data?.data?.map(row => {
-            return {
-              ...row,
-              id: rowWhere(omit(row, ['HyperDXEventType'])),
-            };
-          })
-        : undefined,
-    [data, rowWhere],
-  );
+  const isFetching = traceIsFetching || logIsFetching;
+  const rows = [...traceRowsData, ...logRowsData];
+
+  rows.sort((a, b) => {
+    const aDate = TimestampNano.fromString(a.Timestamp);
+    const bDate = TimestampNano.fromString(b.Timestamp);
+    const secDiff = aDate.getTimeT() - bDate.getTimeT();
+    if (secDiff === 0) {
+      return aDate.getNano() - bDate.getNano();
+    } else {
+      return secDiff;
+    }
+  });
 
   // 3 Edge-cases
   // 1. No spans, just logs (ex. sampling)
   // 2. Spans, but with missing spans inbetween (ex. missing intermediary spans)
   // 3. Spans, with multiple root nodes (ex. somehow disjoint traces fe/be)
 
-  const spanIds = useMemo(() => {
-    return new Set(
-      rows
-        ?.filter(result => result.HyperDXEventType === 'span')
-        .map(result => result.SpanId) ?? [],
-    );
-  }, [rows]);
-
   // Parse out a DAG of spans
-  type Node = SpanRow & { id: string; children: SpanRow[] };
+  type Node = SpanRow & { id: string; parentId: string; children: SpanRow[] };
   const rootNodes: Node[] = [];
-  const nodes: { [SpanId: string]: any } = {};
+  const parentSpanIdsMap = new Map();
+  const nodesMap = new Map();
+  for (const { ParentSpanId, SpanId } of rows ?? []) {
+    if (ParentSpanId && SpanId) parentSpanIdsMap.set(SpanId, ParentSpanId);
+  }
+  let logSpanCount = -1;
   for (const result of rows ?? []) {
+    const { type, HyperDXEventType, SpanId } = result;
+    // ignore everything without spanId
+    if (!SpanId) continue;
+    if (type === SourceKind.Log) logSpanCount += 1;
+
+    // log have dupelicate span id, tag it with -log-couunt
+    const nodeSpanId =
```

Comment:
is this id used anywhere? is it effectively just random and never used further?

Reply:
It uses as ref key to get node, it only use for trace event. And for log event, I add `-log-count` to prevent it overwrite trace event node.

And the reason is, in trace table, all `spanId` is unique, but in log table, `spanId` can be duplicate and it is a reference to trace event(more like `parentSpanId`).

Reply:
Right but in this case we'll never actually use this id in a get call right? Since this logic isn't ported anywhere else and the ID isn't deterministic on the row I can't imagine we're actually using this anywhere else?

Basically we can do `nodeSpanId = Math.random()` and the code will still work as-is? If so, I'm wondering if it's worth keeping all this counting logic around and just substituting it with a random ID and commenting that for context.

This is probably a nit fwiw - just find it confusing to track this code if the logic isn't being used.

Reply:
True, originally I just want to keep them in reference, so we can quick ref it in future if we have to do anything with it.

However since it makes people confuse reading the code. I will just remove it, we can add back if we need it.

Reply:
thanks, do we still need a a random identifier appended or something to the log so that multiple logs for the same span don't get wiped out?

Reply:
nah, because log can only be child, only parent(trace span) have a chance to get ref.

Reply:
Oh got it, logs are never set into `nodesMap`... in that case the id can literally be blank string and not matter?

Reply:
yes, you are right.